src/QsCompiler/CompilerOptimizations/VariableRenaming.fs

```diff
@@ -0,0 +1,124 @@
+    member internal this.getNumUses name =
+        Map.tryFind name numUses
+
+
+    override scope.Transform x =
+        if skipScope then
+            skipScope <- false
+            base.Transform x
+        else
+            constants <- enterScope constants
+            let result = base.Transform x
+            constants <- exitScope constants
+            result
+
+    override scope.Expression = { new ExpressionTransformation() with
+        override expr.Kind = { new ExpressionKindTransformation() with
+            override exprKind.ExpressionTransformation ex = expr.Transform ex
+            override exprKind.TypeTransformation t = expr.Type.Transform t
+
+            override exprKind.onIdentifier (sym, tArgs) =
+                maybe {
+                    let! name =
+                        match sym with
+                        | LocalVariable name -> Some name.Value
+                        | _ -> None
+                    let! newName = tryGetVar constants name
+                    numUses <- Map.add newName (Map.find newName numUses + 1) numUses
+                    return Identifier (LocalVariable (NonNullable<_>.New newName), tArgs)
+                } |? Identifier (sym, tArgs)
+        }
+    }
+
+    override scope.StatementKind = { new StatementKindTransformation() with
+        override stmtKind.ExpressionTransformation x = scope.Expression.Transform x
+        override stmtKind.LocationTransformation x = scope.onLocation x
+        override stmtKind.ScopeTransformation x = scope.Transform x
+        override stmtKind.TypeTransformation x = scope.Expression.Type.Transform x
+
+        override this.onSymbolTuple syms =
+            match syms with
+            | VariableName item -> VariableName (NonNullable<_>.New (generateUniqueName item.Value))
+            | VariableNameTuple items -> Seq.map this.onSymbolTuple items |> ImmutableArray.CreateRange |> VariableNameTuple
+            | InvalidItem | DiscardedItem -> syms
+
+        override this.onVariableDeclaration stm =
+            let rhs = this.ExpressionTransformation stm.Rhs
+            let lhs = this.onSymbolTuple stm.Lhs
+            QsBinding<TypedExpression>.New stm.Kind (lhs, rhs) |> QsVariableDeclaration
+
+        override this.onForStatement stm =
+            let iterVals = this.ExpressionTransformation stm.IterationValues
+            let loopVar = fst stm.LoopItem |> this.onSymbolTuple
+            let loopVarType = this.TypeTransformation (snd stm.LoopItem)
+            let body = this.ScopeTransformation stm.Body
+            QsForStatement.New ((loopVar, loopVarType), iterVals, body) |> QsForStatement
+
+        override this.onRepeatStatement stm =
+            constants <- enterScope constants
+            skipScope <- true
+            let result = base.onRepeatStatement stm
+            constants <- exitScope constants
+            result
+
+        override this.onQubitScope (stm : QsQubitScope) =
```

Comment:
and same question here

Reply:
Ok, I think I get it: you have written this basically specifically to remove shadowing - i.e. this transformation only does something if you actually have shadowed variables due to a prior inlining. It seems a bit of a risk that 1) you actually have to overload all statements that do any kind of binding and 2) you are duplicating the logic regarding the life times of variables. I am not sure it's worth trying to minimize the renaming in this way. What are your thoughts here.